### DISCUSSION TITLE
RelaxRandomVariablesGenerator

### DISCUSSION THREAD

**Discussion by C8504** [google/or-tools#3361] [CP-SAT questions] [answered]
Hi Experts,

As I know, this generator didn't specify variable types.

However, variables are grouped by specified class,  binary variables VS integer variables, or variables related resource VS variables related other resources, more like, variables indexed with day VS variables indexed with other days.

So, could users control which kind variables to be relaxed in Neighborhood？
like, fixing 0-1 variables only to improve.

I guess this maybe callback feature for LNS solvers in CP-SAT, how do you think?

Thanks.
John

**Comment by lperron** [comment, marked as answer]
there are lots of generators. This one selects a random subset of variables. It does exactly what it is meant to do.

Callback for LNS is hard because of presolve. A lot of variables disappear, are rescaled, during presolve.

**Reply by C8504** [reply]
ok, understood. thanks lperron.
